### ISSUE TITLE
sf!

### ISSUE THREAD

**Issue by mpadge** [osmdatar/osmdata#17]
The whole package is currently structured around `sp` objects, with `overpass_query` returning an `osmdata` object embedded within which are three `sp` objects (`points/lines/polygons`). How can `sf`-class objects best be incorporated? Possibilities:

1. Have `overpass_query` return some more generic object the members of which can be accessed via:

   (i) `sp_osm_points()`, `sp_osm_lines()`, `sp_osm_polygons()`; and 

   (ii) `sf_osm_points()`, `sf_osm_lines()`, `sf_osm_polygons()` 

2. Keep things as they are and include an `osmdata_to_sf` function which will convert all `sp` members to `sf`, either by:

   (i) Simply including an `sf` dep - a massive increase in package size and complexity!, or

   (ii) Hard-coding `sf` construction within the `Rcpp` routines as currently done for `sp`

Thoughts @Robinlovelace ?

**Comment by Robinlovelace** [comment]
I think 2. ii) is a decent starting point but not sure how it's different than saying, e.g. in the vignette, something like:

> Currently **osmdata** outputs `Spatial` objects, as defined by the **sp** package. If you would like to convert them to other spatial classes, use external packages - an example is shown below for **sf**:

```r
library(osmdata)
b = tmap::bb("Leeds")
q = add_feature(opq(b), "amenity", "pub")
p = overpass_query(q)
mapview::mapview(p$osm_points) # check the results
psf = sf::st_as_sf(p$osm_points)
head(psf)
```


Then as **sf** becomes more mature we could eventually include it as an import/dependency due to other benefits of using it.

I see it as no different form importing sp as we currently do and think it's good to be an early adopter.

**Comment by mpadge** [comment]
Just put up the first rough comparison of `osmdata` and `sf` in a new [wiki page](https://github.com/osmdatar/osmdata/wiki/sf-comparisons) - `GDAL` output as interpreted by `sf` actually differs in quite categorical ways, notably through, for example, `multipolygons` including all points defining those geometries, with these points then **not** included in any other layers they are not part of. In comparison, `osmdata` returns a points object containing every single point in the OSM data.

More concretely for the example given in the wiki:

* `osmdata::overpass_query ()$points` return 5,071 points
* `nrow (sf::st_read (file, layer='points')) = 137` - the number of unique point objects

The two approaches are doing completely different things. Current issue as I see it:

* I like the way `osmdata` does it at present, because it seems to much more directly reflect the way that OSM data are actually structured; however
* `GDAL` is and ought to be seen as a standard, and we are clearly operating in fairly grave violation of that.

Reconciliation would require quite a comprehensive restructure of the `Rcpp` code to more directly reflect what `GDAL` does. Note, however, one big disadvantage of the `GDAL/sf` approach for OSM purposes:

* Sticking all points within, for example, the `multipolygon` geom removes any ability to recover any OSM values for those points. The construction of a `multipolygon` (or any other) geometry implicitly assumes that geometry to represent the sole level of interest, so full info will be retained for each object in the multipolygon list, while any and all info regarding the components that make up each multipolygon---including all `key-val` pairs from underlying lines or points---simply gets discarded.
* I'm not sure this is the kind of standard the typical person interested in OSM data would be happy with. OSM has been intentionally structured to allow tracing ''down'' to discern the properties of any and all components making up a given feature - hence the ubiquitous [overpass ''>'' operator](http://wiki.openstreetmap.org/wiki/Overpass_API/Overpass_QL#Recurse_down_.28.3E.29) which [we also use](https://github.com/osmdatar/osmdata/blob/master/R/opq.R#L74).
* In short: The `GDAL/sf` approach removes an ability to subsequently recurse down or up. I'd suspect that the amount of effort involved in conceiving of and implementing the `overpass` recursion operators reflects the abiding importance of this aspect of OSM data structure to those using such data.
* This suggests to me that `sf` is an inappropriate framework for OSM data?

**Comment by Robinlovelace** [comment]
Wow - very interesting findings and an excellent description of what is going on with OSM data and why it doesn't fit neatly into the straitjacket of conventional spatial classes, in QGIS or R. 

I agree that this in depth description shows another reason why osmdata is needed. It also explains the use of functions such as `find_down()` in [**osmar**](https://cran.rstudio.com/web/packages/osmar/index.html), as described in this tutorial I wrote way back in the days when that was cutting edge (it's still worth benchmarking against and feature comparing - happy to do some work on that): https://github.com/Robinlovelace/osm-tutorial/blob/master/osm.md

**Comment by mpadge** [comment]
Oh yeah, I'd forgotten that `osmar` had those `find_down()` type capabilities. I always found that a bit confusing because it was never clear *what* one was meant to be finding. Conversely, I'd suggest my suggestion in #18 offers (and indeed requires) pretty unambiguous clarity in what is to be recursively found. 

Want points that make up polygons? Simply ask for `points (polygons)`
Want all polygons that include a bunch of points? Simply ask for `polygons (points)`

The thing you want at the end is what you ask for, without even needing to think about down and up.

**Comment by Robinlovelace** [comment]
Agreed.

**Comment by mpadge** [comment]
So [this commit](https://github.com/osmdatar/osmdata/commit/7cd3421f2e9d1d291f6e9e806b43408e23116ec9) implements what I think will provide a robust framework for `sf` implementation. It codes in straight `R` code what will need to be implemented within the `C++` functions to spit out `sf` output. The point of these tests is to always ensure that the output of `osmdata` will always be consistent with `sf` output. @Robinlovelace Your initial thoughts would be very useful before I proceed any further on this ...

**Comment by mpadge** [comment]
`sf` implementation done! i'm sure it ain't yet perfect, but problems can be dealt with in separate issues, so closing this now. See also #20 and #14.

(Note that only `osmdata_xml` and `osmdata_sf` work at present - `osmdata_sp` is likely broken until I fix it to reflect `osmdata_sf`.)

**Comment by mpadge** [comment]
... already seen that several of the README examples don't work, but i also added an initial example that does work. Re-opened temporarily until all examples work ...

**Comment by mpadge** [comment]
All examples work now, so closing properly this time.
